### PR TITLE
[SPARK-41294][SQL] Assign a name to the error class _LEGACY_ERROR_TEMP_1203 / 1168

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
In the PR, I propose to assign the name `NUM_COLUMNS_MISMATCH` to the error class `_LEGACY_ERROR_TEMP_1203` and `_LEGACY_ERROR_TEMP_1168`.

### Why are the changes needed?
Proper names of error classes should improve user experience with Spark SQL.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @MaxGekk

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -876,6 +876,13 @@
     ],
     "sqlState" : "42000"
   },
+  "NOT_ENOUGH_DATA_COLUMNS" : {
+    "message" : [
+      "Cannot write to <tableName>, not enough data columns:",
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:881]: The condition is more strict:
https://github.com/apache/spark/blob/a80899f8bef84bb471606f0bbef889aa5e81628b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala#L380

it checks exact number of columns but not `enough data columns:`.

**Review by MaxGekk — COMMENTED** [review]
cc @srielau

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:881]: Can we re-use the error class `NUM_COLUMNS_MISMATCH` in the case? If not, would it possible to make it more generic to cover the case too.

Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:881]: Let me try

Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:881]: @MaxGekk Can I change the template of `NUM_COLUMNS_MISMATCH ` error message to "<operator> can only be performed on tables with the same number of columns, but the first table (<tableName>)'s columns(length: <refLength>): [refColumns], the (ordinal: <invalidOrdinalNum>)'s columns(length: <invalidLength>): [invalidColumns]." ?

The original is "<operator> can only be performed on tables with the same number of columns, but the first table has <refNumColumns> columns and the <invalidOrdinalNum> table has <invalidNumColumns> columns."

Comment (srielau) [reply core/src/main/resources/error/error-classes.json:881]: Given that we have columns defaults, at least in the case of insert it's not really about the target table, but the column list.
I take it the same error is raised on UPDATE T SET  (...) = (...)?  and UNION/EXCEPT/...  ( ) .. IN ( SELECT ...) ?

How about <operator> expects matching number of columns. But the left side has <leftNumCols> while the right side has <rightNumCols>.

Alternatively we could talk about "assignments" using a different error class (?)

Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:881]: I will open a new pr to make the error class NUM_COLUMNS_MISMATCH more generic.

Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:881]: @panbingkun @srielau 
New Pr: https://github.com/apache/spark/pull/38937

After above, I will continue this pr.

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:881]: @panbingkun The PR https://github.com/apache/spark/pull/38937 has been merged already. Please, continue with this.

Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:881]: OK, Let me continue to do it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -2018,15 +2007,29 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
-  def cannotWriteNotEnoughColumnsToTableError(
-      tableName: String, expected: Seq[Attribute], query: LogicalPlan): Throwable = {
     new AnalysisException(
-      errorClass = "_LEGACY_ERROR_TEMP_1203",
+      errorClass = "NUM_COLUMNS_MISMATCH",
       messageParameters = Map(
-        "tableName" -> tableName,
-        "tableColumns" -> expected.map(c => s"'${c.name}'").mkString(", "),
-        "dataColumns" -> query.output.map(c => s"'${c.name}'").mkString(", ")))
+        "operator" -> toSQLStmt(operator.nodeName),
```
Comment (panbingkun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2027]: The value here may be: INSERTINTOSTATEMENT, APPENDDATA.
It's weird, maybe we should pin it to `INSERT INTO`?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2027]: Maybe, we should introduce `prettyName()` similar to `prettyName` in `Expression`. @cloud-fan @HyukjinKwon WDYT?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2027]: > It's weird, maybe we should pin it to INSERT INTO?

So far, let's do that.

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -904,6 +904,13 @@
     ],
     "sqlState" : "42000"
   },
+  "NOT_ENOUGH_DATA_COLUMNS" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:907]: Is it used somewhere? if not, please, remove it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 4]
```diff
@@ -2019,14 +2008,27 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
...
     new AnalysisException(
-      errorClass = "_LEGACY_ERROR_TEMP_1203",
+      errorClass = "NUM_COLUMNS_MISMATCH",
       messageParameters = Map(
-        "tableName" -> tableName,
-        "tableColumns" -> expected.map(c => s"'${c.name}'").mkString(", "),
-        "dataColumns" -> query.output.map(c => s"'${c.name}'").mkString(", ")))
+        "operator" -> "INSERT INTO",
+        "firstNumColumns" -> firstNumColumns.toString,
+        "invalidOrdinalNum" -> ordinalNumber(invalidOrdinalNum + 1),
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2028]: To get `first`, `invalidOrdinalNum` should be -1?

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala]: This is really weird. Let me modify it more reasonably.

File: core/src/main/resources/error/error-classes.json (hunk 3 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: Done.

File: sql/core/src/test/resources/sql-tests/results/postgreSQL/numeric.sql.out [hunk 5]
```diff
@@ -3831,12 +3831,12 @@ struct<>
...
-  "errorClass" : "_LEGACY_ERROR_TEMP_1168",
+  "errorClass" : "NUM_COLUMNS_MISMATCH",
   "messageParameters" : {
-    "insertedColumns" : "5",
-    "staticPartCols" : "0",
-    "tableName" : "`spark_catalog`.`default`.`num_result`",
-    "targetColumns" : "3"
+    "firstNumColumns" : "3",
+    "invalidNumColumns" : "5",
+    "invalidOrdinalNum" : "second",
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/postgreSQL/numeric.sql.out:3838]: @cloud-fan @srielau Are you ok that we are talking about inputs (and their ordinal number), but not a table name.

Comment (srielau) [reply sql/core/src/test/resources/sql-tests/results/postgreSQL/numeric.sql.out:3838]: Not really Ok with that. I'm also confused by the "second" portion. That doesn't translate and seems unnecessary compared to "1." and "2.".
Should we have a dedicated error message? Especially INSERT INTO has an optional column list, so there are many ways to "fix" the error. It's not the same as a UNION ALL or an EXCEPT.

Comment (itholic) [reply sql/core/src/test/resources/sql-tests/results/postgreSQL/numeric.sql.out:3838]: @panbingkun Any update on this? Can we have this on Apache Spark 3.4 ?

**Pull request closed without merging**